### PR TITLE
Improve PE File & Program Icon Loading

### PR DESCRIPTION
**Pull request by divadretlaw**
## PEFile Improvements

- Close `FileHandle` after parsing
- Split types into multiple files
- Add some documentation
- Fix `OptionalHeader` not differentiating PE32 and PE32+
- Interpret more values while parsing (maybe a bit overkill, but they could be useful in the future or other projects trying to access values of a PE file)
- Improve Icon Loading

I think it would be best to cache the icons as it seems to take a bit to load them to find the best / largest one.

I would recommend https://github.com/kean/Nuke but there are many options and even the possibility to create a cache without any dependencies if that is desired.

Adding a dependency should be the decision of the maintainer and if they think adding a cache, and also which library to use if any, then I would be happy to implement that as well.

### CODE REVIEW

**Comment by JoshuaBrest** [comment]
What do you mean by cache?

**Comment by IsaacMarovitz** [comment]
Most of the improvements to PE parsing that I'm seeing here are not relevant. The parsing implementation used was designed to be a bit naive, as we really only need the data present in the resource table for our purposes. If we wanted to create a more complete implementation, as is shown here, it might be worth splitting it into a separate project/swift package.

**Comment by divadretlaw** [comment]
That approach makes sense, I reverted the not-needed changes to the PE parsing

File: WhiskyKit/Sources/WhiskyKit/Extensions/FileHandle+Extensions.swift [hunk 1]
```diff
@@ -20,6 +20,19 @@ import Foundation
 import os.log
 
 extension FileHandle {
+    func extract<T>(_ type: T.Type, offset: UInt64 = 0) -> T? {
```
Comment (IsaacMarovitz) [inline WhiskyKit/Sources/WhiskyKit/Extensions/FileHandle+Extensions.swift:23]: Is the `Data` extract extension still used? If not I'd remove it.

File: Whisky/Views/Bottle/Pins/PinView.swift [hunk 2]
```diff
@@ -78,13 +78,14 @@ struct PinView: View {
...
         }
-        .onAppear {
+        .task {
             name = pin.name
-            Task.detached { @MainActor in
-                if let peFile = program.peFile {
-                    image = peFile.bestIcon()
-                }
+            guard let peFile = program.peFile else { return }
+            let task = Task<Image?, Never>.detached {
```
Comment (IsaacMarovitz) [inline Whisky/Views/Bottle/Pins/PinView.swift:84]: Would removing `@MainActor` here not be sufficient? This seems a little verbose.

**Review by IsaacMarovitz — CHANGES_REQUESTED** [review]

File: WhiskyKit/Sources/WhiskyKit/PE/PortableExecutable.swift [hunk 3]
```diff
@@ -17,291 +17,176 @@
...
+        }
+        defer {
+            try? handle.close()
         }
-        coffFileHeader = try COFFFileHeader(handle: handle)
+
+        return rsrc(handle: handle)
     }
 
+    #if canImport(AppKit)
```
[the remark this answers is not in the document]
Comment (IsaacMarovitz) [reply WhiskyKit/Sources/WhiskyKit/PE/PortableExecutable.swift:148]: https://developer.apple.com/documentation/coregraphics/cgcontext/1455939-init

**Review by IsaacMarovitz — APPROVED** [review]

**Pull request merged**